File: src/com/leozhi/topic861/Solution03.java
package com.leozhi.topic861;

/**
 * @author leozhi
 * 贪心算法
 * 官方题解
 * 通过
 * 0ms
 */
public class Solution03 {
    public int matrixScore(int[][] A) {
        int m = A.length, n = A[0].length;
        // 最高位全为 1，即 1 左移 n - 1 位
        int res = m * (1 << (n - 1));
        for (int i = 1; i < n; i++) {
            int nOnes = 0;
            for (int[] ints : A) {
                if (ints[0] == 1) {
                    nOnes += ints[i];
                } else {
                    nOnes += (1 - ints[i]);
                }
            }
            int j = Math.max(nOnes, m - nOnes);
            res += j * (1 << (n - i - 1));
        }
        return res;
    }
}

File: src/com/leozhi/topic474/Solution03.kt
package com.leozhi.topic474

import kotlin.math.max

/**
 * @author leozhi
 * 动态规划
 * 通过
 * 228ms
 */
class Solution03 {
    fun findMaxForm(strs: Array<String>, m: Int, n: Int): Int {
        val dp = Array(m + 1) { IntArray(n + 1) }
        for (str in strs) {
            val zeroCount = str.count { it == '0' }
            val oneCount = str.length - zeroCount
            for (i in m downTo zeroCount) {
                for (j in n downTo oneCount) {
                    dp[i][j] = max(dp[i][j], dp[i - zeroCount][j - oneCount] + 1)
                }
            }
        }
        return dp.last().last()
    }
}